vtex-apps/checkout
# Fix payment step crash due to responsive-layout

#### What problem is this solving?

This PR fixes an issue in the Payment step that occurred when you resized the window from desktop -> mobile -> desktop. The error is because of a library in the `CreditCard` component inside the payment which is being used to communicate with the credit card `iframe` based on events built on top of `window.postMessage`.

The library has a restriction that only one event must exist at a time (e.g. you can't listen to, say, `paymentSystems` twice) and _apparently_ our `CreditCard` component was being rendered twice 😱 which caused us to listen to the same event as the other component instance, so the library throwed an error when the second instance begin rendering.

So, it turns out that the issue wasn't with our component, but with the `responsive-layout` component from the SF team. This layout component is built on top of the `device-detector` app, and each block of this app (`responsive-layout.desktop`, `responsive-layout.mobile`, etc) used a _separate_ "instance" of the hook exposed in `device-detector`. So, when you changed the resolution of the page, each component received its own event to update itself causing for a brief moment for both components we use (desktop and mobile) to be mounted in the page (see screenshot below), and causing the error I described above.

![image](https://user-images.githubusercontent.com/10223856/81412189-df03ce80-9119-11ea-922d-41f449dcba47.png)

#### How should this be manually tested?

[Workspace](https://lucas--checkoutio.myvtex.com/cart/add?sku=289). You can see the issue in [this workspace](https://chkio--checkoutio.myvtex.com/cart/add?sku=289).

Test plan:

1. Proceed to checkout and fill all informations until the payment step
2. Resize the screen from desktop to mobile, then to desktop again.
3. The page shouldn't crash.

#### Checklist/Reminders


#### Screenshots or example usage

![payment-resize](https://user-images.githubusercontent.com/10223856/81414984-2db36780-911e-11ea-9608-d223c1ec482d.gif)

#### Type of changes


✔️ | Type of Change
---|---
✔️ | Bug fix 
_ | New feature 
_ | Breaking change 
_ | Technical improvements
